km/km_filesys.c

```diff
@@ -2077,41 +2077,76 @@ static inline int do_full_write(int fd, char* bufp, size_t bufl)
  *    We know the buffer is big enough because the buffer for all of the
  *    elf notes is precomputed before we build the notes.
  *  length - amount of space available at what buf points to.
- *  file - the fd we can read the buffered pipe data from.
- *  writefd - the fd we need to write the data back into so that it will
- *    be there for the payload to read if we resume after taking a snapshot.
+ *  fd - the fd we can read the buffered pipe data from.
  *  queuedbytes - the number of bytes queued in the pipe/socketpair.
  * Returns:
- *  The number of bytes written into the elf note.
+ *  0 - success
+ *  != 0 - errno for the failure
  */
-static inline int
-fs_core_save_pipe_contents(char* buf, size_t length, km_file_t* file, int writefd, size_t queuedbytes)
+static inline int fs_core_save_pipe_contents(char* buf, size_t length, int fd, size_t queuedbytes)
 {
+   km_file_t* file = &km_fs()->guest_files[fd];
    km_assert(queuedbytes <= length);
+   km_assert(fd >= 0);
 
-   // Read the bytes queued in the pipe
-   int rc = do_full_read(file->ofd, buf, queuedbytes);
+   // Read the bytes queued in the pipe into the memory buf points to
+   int rc = do_full_read(fd, buf, queuedbytes);
    if (rc != 0) {
       return rc;
    }
 
-   // now write the bytes back into the pipe
-   return do_full_write(writefd, buf, queuedbytes);
+   if (file->ofd < 0) {
+      // The write side of this pipe is closed.  So we need to recreate the pipe
+      // to put the data back into the pipe buffer and then close the write end
+      // after writing the data.
+      // If any of this fails, we can not allow the payload to resume since we can't
+      // recreate is half open pipe or sockpair.
+      int fdpair[2];
+      if (file->how == KM_FILE_HOW_SOCKETPAIR0 || file->how == KM_FILE_HOW_SOCKETPAIR1) {
+         if (socketpair(file->sockinfo->domain, file->sockinfo->type, file->sockinfo->protocol, fdpair) <
+             0) {
+            km_err(1, "Unable to recreate half open socketpair, fd %d", fd);
```

Comment:
should this be warn and return an error?

Reply:
No.  We have read data from a half open socketpair.  Now that we have read the data we need to recreate the socketpair, write the data back into it, and then close the end we've written the data into so that the payload will have the same half open socketpair it had before the snapshot was requested.  If any of these steps fail,  there is not much we can do but terminate the payload.

Reply:
Maybe you have another way to solve this?